### ISSUE TITLE
[1.3 AMI] Quickly restarting scylla-server after a restart makes scylla to dump core

### ISSUE THREAD

**Issue by lmr** [scylladb/scylladb#1672]
_Installation details_
Scylla version (or git commit hash): 1.3 AMI

When scylla-server is killed, systemd is supposed to restart it. If you issue a restart command a few seconds later, you'll trigger a coredump, whose backtrace is summarized here:

```
PID: 14064 (scylla)
UID: 994 (scylla)
GID: 1001 (scylla)
Signal: 6 (ABRT)
Timestamp: Fri 2016-09-09 04:24:44 UTC (4 days ago)
Command Line: /usr/bin/scylla --log-to-syslog 1 --log-to-stdout 0
--default-log-level info --collectd-address=127.0.0.1:25826
--collectd=1 --collectd-poll-period 3000 --network-sta
Executable: /usr/bin/scylla
Control Group: /system.slice/scylla-server.service
Unit: scylla-server.service
Slice: system.slice
Boot ID: 33dfb6109c6049b3a49669f3593cafdd
Machine ID: 8ea3e6ea03b64d018847fcdf669935d4
Hostname: ip-172-31-34-206
Coredump: /var/lib/systemd/coredump/core.scylla.994.33dfb6109c6049b3a49669f3593cafdd.14064.1473395084000000
Message: Process 14064 (scylla) of user 994 dumped core.

     Stack trace of thread 14064:
     #0  0x00007f4d440095f7 raise (libc.so.6)
     #1  0x00007f4d4400ace8 abort (libc.so.6)
     #2  0x00007f4d44002566 __assert_fail_base (libc.so.6)
     #3  0x00007f4d44002612 __assert_fail (libc.so.6)
     #4  0x0000000000465044
_ZN7reactor7at_exitESt8functionIF6futureIJEEvEE (scylla)
     #5  0x00000000004e0ddb
_ZN7reactor7at_exitESt8functionIF6futureIJEEvEE (scylla)
     #6  0x000000000072c29f _ZZZ4mainENKUlvE_clEvENKUlvE_clEv (scylla)
     #7  0x000000000073a3b6 apply (scylla)
     #8  0x000000000047131e _ZNKSt8functionIFvvEEclEv (scylla)
     #9  0x0000000000581de2 _ZN7seastar14thread_context6s_mainEjj (scylla)
     #10 0x00007f4d4401b110 n/a (libc.so.6)
```

This is a known issue, see the e-mail thread

https://groups.google.com/forum/#!topic/seastar-dev/xL8HqaX9R4o

This issue was opened to track its status

**Comment by tgrabiec** [comment]
Solution could be to use cooperative interruption/shutdown and get rid of `at_exit()`.

SIGINT handler would not initiate reactor shutdown, but would "interrupt" reactors (only once). Interrupting would set "interrupted" flag on the reactor. 

Interruption would cause some operations, which belong to "interruptible" class, to resolve immediately with interrupted_exception when reactor is interrupted. For example `sleep()` could be interruptible. It would resolve with an exception, interrupting the normal flow. The exception interrupts the normal execution flow, and can be handled. When it propagates out of the main application loop, it triggers shutdown sequence (see https://github.com/scylladb/seastar/commit/9267dfa3f72271c05d59a37d8903855884632d01 for an example).

The application would look like this:

```
app_template::run([] {
   return seastar::thread([] {
       auto x = start_x().get0();  // <-- can throw interrupted_exception
       auto stop_x = defer([] { x.stop().get(); });
       auto y = start_y();  // <-- can throw interrupted_exception
       auto stop_y = defer([] { x.stop().get(); });
       sleep_forever().get(); // <-- can throw interrupted_exception
   });
});
```

**Comment by gleb-cloudius** [comment]
On Thu, Sep 15, 2016 at 05:33:06AM -0700, Tomasz Grabiec wrote:
> Solution could be to use cooperative interruption/shutdown and get rid of `at_exit()`.
> 
> SIGINT handler would not initiate reactor shutdown, but would "interrupt" reactors (only once). Interrupting would set "interrupted" flag on the reactor. 
> 
> Interruption would cause some operations, which belong to "interruptible" class, to resolve immediately with interrupted_exception when reactor is interrupted. For example `sleep()` could be interruptible.
There is already interruptible sleep: sleep_abortable(). I wanted to
make default sleep() abortable but there was resistance, but the truth
is if we want ^C to work reliable there is no place for non abortable
sleep.
 
> It would resolve with an exception, interrupting the normal flow. The exception interrupts the normal execution flow, and can be handled. When it propagates out of the main application loop, it triggers shutdown sequence (see https://github.com/scylladb/seastar/commit/9267dfa3f72271c05d59a37d8903855884632d01 for an example).
We need to audit all the places that do catch(...) since they will "eat"
exception and stop shutdown process. Actually we will have to audit all
places since I am sure there are places that do not expect to get
exception and will result in "shutdown exception" to be ignored.


> 
> The application would look like this:
> 
> ```
> app_template::run([] {
>    return seastar::thread([] {
>        auto x = start_x().get0();  // <-- can throw interrupted_exception
>        auto stop_x = defer([] { x.stop().get(); });
>        auto y = start_y();  // <-- can throw interrupted_exception
>        auto stop_y = defer([] { x.stop().get(); });
>        sleep_forever().get(); // <-- can throw interrupted_exception
>    });
> });
> ```
> 
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/scylladb/scylla/issues/1672#issuecomment-247315253

--
			Gleb.

**Comment by nyh** [comment]
Unfortunately, I can still reproduce this problem on today's master (post 3.0), and it isn't specific to version 1.3, to AMI, or to the scenario described above (so maybe the issue should be renamed).

The underlying problem is that if we gently kill (with control-C aka SIGINT, or SIGTERM) Scylla while it is still booting, it will often crash in its attempt to shut down.

In today's master, I see crashes like
```
scylla: src/core/reactor.cc:2545: void seastar::reactor::at_exit(std::function<seastar::future<>()>): Assertion `!_stopping' failed.
Aborting on shard 0.
Backtrace:
  0x0000000000abdd22
  0x00000000009ecca5
  0x00000000009ecfa5
  0x00000000009ed053
  /lib64/libpthread.so.0+0x00000000000120bf
  /lib64/libc.so.6+0x0000000000036eaa
  /lib64/libc.so.6+0x00000000000215b8
  /lib64/libc.so.6+0x0000000000021490
  /lib64/libc.so.6+0x000000000002f611
  0x00000000009c7f5c
  0x00000000009f18c7
  0x000000000101b058
  0x0000000001021d6a
  0x000000000095b2b5
```
Unsurprisingly (we could have guessed this from the error message), this decodes to:
```
__GI___assert_fail at :?
sigabrt_action at /home/nyh/scylla/seastar/src/core/reactor.cc:4029
 (inlined by) operator() at /home/nyh/scylla/seastar/src/core/reactor.cc:4011
 (inlined by) _FUN at /home/nyh/scylla/seastar/src/core/reactor.cc:4007
seastar::reactor::at_exit(std::function<seastar::future<> ()>) at /home/nyh/scylla/seastar/src/core/reactor.cc:2545
seastar::reactor::at_exit(std::function<seastar::future<> ()>) at /home/nyh/scylla/seastar/src/core/reactor.cc:2544 (discriminator 1)
```

So basically, after Scylla was already shut down and ran reactor::run_exit_tasks() which sets reactor::_stopping = true, Scylla ran at_exit() - to register new exit tasks - and that is forbidden.

It is quite easy to reproduce this bug - just start one node of Scylla, and then "kill" it (with SIGTERM or SIGINT - e.g., manual control-C) after a varying duration - from a fraction of a second until 10 seconds. However, it doesn't happen all the time (sometimes I see this almost every time, sometimes I can run a few runs without a crash).

**Comment by denesb** [comment]
Closing as stale.
